Adds option to not become when generating keys locally

nkakouros opened the pull request:
When generating keys, the role runs the key generation tasks with `connection: local`. However, most probably the role will be run with `become: true` to be able to edit users' ssh-related files on the remote hosts. But in this case the local tasks will also run as root. This might not be what the user wants. For instance, `sudo` may require a password locally and this will break ansible if no become_password is provided.

tersmitten commented on README.md:
```diff
@@ -49,6 +49,8 @@ None
 * `ssh_keys_known_hosts.{n}.enctype`: [required]: The type of the fingerprint
 * `ssh_keys_known_hosts.{n}.fingerprint`: [required]: The actual fingerprint
 
+* `ssh_keys_generate_local_become`: [optional, default: false]: Whether to use sudo when generating ssh keys locally
```
Can you change it to ``ssh_keys_generate_keys_local_become: [default: `false`]: ``?

nkakouros replied:
Done

tersmitten commented on tasks/generate.yml:
```diff
@@ -45,3 +45,4 @@
       tags:
         - ssh-keys-generate-public-keys
   connection: local
+  become: "{{ ssh_keys_generate_local_become }}"
```
What do you think about?

become: "{{ ssh_keys_generate_keys_local_become | default(omit) }}"

nkakouros commented:
@tersmitten I rebased and addressed your comment. I added one more commit. I believe this should also be included. The problem with `connection: local` is that it runs with the "environment" of the remote host. This includes variables like `ansible_python_interpreter` (if it is set in the inventory for the remote host). This is a problem because the remote python interpreter path may not coincide with the local one, as is the case when running ansible in a virtual environment. `delegate_to` avoids this issue.

nkakouros commented:
The CI errors are, I think, not related to this PR.

tersmitten commented:
Can you sync with `master`?

nkakouros commented:
Done.

nkakouros commented:
@tersmitten Gentle ping for when you have time.

tersmitten commented:
I'll try to have a look upcoming week

tersmitten commented:
After reading [this](http://willthames.github.io/2018/07/01/connection-local-vs-delegate_to-localhost.html) and [this](https://docs.ansible.com/ansible/latest/user_guide/playbooks_delegation.html) I'm not if this is the way to go.

nkakouros commented:
Your objection has to do with the second commit and the replacing of `connection: local` with `delegate_to: localhost` right?

The issue boils down to whether one prefers `connection: local` or `delegate_to: localhost`. The latter in my opinion is the most trouble free. I can't see where it will fail. The former needs to setup the python interpreter. The problem is that it is not easy to do this if you need to setup the interpreter for the remote host also. E.g., if you have:

```yaml
# inventory/groups/linux
ansible_python_interpeter: /path/to/specially/compiled/python3/on/server
```

you will need to override the interpreter for the task that uses `connection: local` since messing with assigning hosts to groups will not work since we are talking about a single task from a role (unless I am missing sth). If the task belongs to a role, you will need to `include_role` with `vars: {'ansible_python_interpeter': ...}` but if the role does not run all of its tasks locally this can't work (neither will using `vars` at the playbook level), unless you include the same role twice with different tags. The other option is for the role to set the interpeter for the local task, but that would require to expose low level details to the role user (e.g. via an `ssh_keys_local_python_interpreter` variable). 

The argument that the first article makes (about having to supply `delegate__to` to each and every task is also moot if you have a role that mixes local and remote tasks (where you will need to supply `connection: local` at the task level too). Also, the `block` keyword allows for DRY-er code. 

I am in favor of the simpler `delegate_to: localhost` since there is no drawback that comes with it, at least from what I can tell and all the above options sound like a lot of unneeded complexity and overhead to just not use `delegate_to`.

tersmitten commented:
Ok, let's give it a try

tersmitten commented:
Merged it! Note that I also renamed some stuff. Make sure to have a look.
